### PR TITLE
Add gMSA domainless support

### PR DESCRIPTION
**Pull request by saikiranakula-amzn**
### Summary
This PR adds the  domainless gMSA support on linux establishing communication with credentials-fetcher daemon and also sharing tickets with application containers

### Implementation details
Added gMSA linux domainless support

### Testing
This was tested through unit tests, as well as manually running the agent on the instance and also verifying the gMSA domainless functionality

New tests cover the changes: 

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/taskresource/credentialspec/credentialspec_linux_test.go [hunk 1]
```diff
@@ -154,7 +161,79 @@ func TestHandleSSMCredentialspecFile(t *testing.T) {
...
+	wg.Wait()
+	close(errorEvents)
+	err := <-errorEvents
+	assert.NoError(t, err)
+
+	cs.CredSpecMap[credentialSpecSSMARN] = expectedKerberosTicketPath
+
+	actualKerberosTicketPath, err := cs.GetTargetMapping(credentialSpecSSMARN)
+	assert.NoError(t, err)
+	assert.Equal(t, expectedKerberosTicketPath, actualKerberosTicketPath)
```
Comment (amogh09) [inline agent/taskresource/credentialspec/credentialspec_linux_test.go:175]: `cs.GetTargetMapping` method just gets the value from `cs.CredSpecMap`  which this test itself is setting on L171.  So, what is being tested here?

Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux_test.go:175]: this is just that map is properly populated

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 2]
```diff
@@ -181,12 +228,64 @@ func (cs *CredentialSpecResource) Create() error {
...
+				return err
+			}
+
+			creds := DomainlessUserCredentials{}
+			if err := json.Unmarshal([]byte(asmSecretData), &creds); err != nil {
+				return err
+			}
+			//set up server connection to communicate with credentials fetcher daemon
+			conn, err := credentialsfetcherclient.GetGrpcClientConnection()
+			seelog.Infof("grpc connection: %v", conn)
```
[the remark this answers is not in the document]
Comment (mythri-garaga) [reply agent/taskresource/credentialspec/credentialspec_linux.go:272]: +1

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 3]
```diff
@@ -226,27 +325,17 @@ func (cs *CredentialSpecResource) handleKerberosTicketCreation() error {
 	return nil
 }
 
-func (cs *CredentialSpecResource) handleCredentialspecFile(credentialSpec string, wg *sync.WaitGroup, errorEvents chan error) {
+func (cs *CredentialSpecResource) handleCredentialspecFile(originalCredentialSpec, credentialSpec string, isDomainlessGmsa bool, wg *sync.WaitGroup, errorEvents chan error) {
```
Comment (mythri-garaga) [inline agent/taskresource/credentialspec/credentialspec_linux.go:328]: super nit: can we rename `credentialSpec ` to `credentialSpecFile`?

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 4]
```diff
@@ -337,37 +425,58 @@ func (cs *CredentialSpecResource) handleSSMCredentialspecFile(originalCredential
...
 	//parse json to extract the service account name and the domain name
-	var credentialSpecSchema CredentialSpecSchema
+	if isDomainlessGmsa {
+		var credentialSpecDomainlessSchema CredentialSpecDomainlessSchema
+		// Unmarshal or Decode the JSON to the interface.
+		err := json.Unmarshal([]byte(credSpecContent), &credentialSpecDomainlessSchema)
 
-	// Unmarshal or Decode the JSON to the interface.
-	err := json.Unmarshal([]byte(credSpecContent), &credentialSpecSchema)
+		if err != nil {
```
Comment (mythri-garaga) [inline agent/taskresource/credentialspec/credentialspec_linux.go:445]: nit: indentation is off

File: agent/taskresource/credentialspec/credentialspec_linux.go (hunk 3 above)
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux.go]: changed variable name

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 5]
```diff
@@ -337,37 +425,58 @@ func (cs *CredentialSpecResource) handleSSMCredentialspecFile(originalCredential
...
+			seelog.Errorf("Error unmarshalling credentialspec data %s", credSpecContent)
+			return
```
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux.go]: returned error from the method

File: agent/taskresource/credentialspec/credentialspec_linux.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux.go]: moved the code block accordingly

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 6]
```diff
@@ -181,12 +228,64 @@ func (cs *CredentialSpecResource) Create() error {
...
+
+			asmSecretData, err := asm.GetSecretFromASM(v.domainlessGmsaUserArn, asmClient)
+			if err != nil {
+				seelog.Errorf("failed to retrieve credentials for domainless gMSA user: %s", err)
+				return err
+			}
+
+			creds := DomainlessUserCredentials{}
+			if err := json.Unmarshal([]byte(asmSecretData), &creds); err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux.go]: updated accordingly

File: agent/taskresource/credentialspec/credentialspec_linux.go [hunk 7]
```diff
@@ -337,37 +433,68 @@ func (cs *CredentialSpecResource) handleSSMCredentialspecFile(originalCredential
...
+		var credentialSpecDomainlessSchema CredentialSpecDomainlessSchema
+		// Unmarshal or Decode the JSON to the interface.
+		err := json.Unmarshal([]byte(credSpecContent), &credentialSpecDomainlessSchema)
 
-	// Unmarshal or Decode the JSON to the interface.
-	err := json.Unmarshal([]byte(credSpecContent), &credentialSpecSchema)
+		if err != nil {
+			reason := "Error unmarshalling credentialspec domainless  " + credSpecContent
+			seelog.Errorf(reason)
+			cs.setTerminalReason(reason)
```
Comment (amogh09) [inline agent/taskresource/credentialspec/credentialspec_linux.go:461]: The terminal reason being set here is overridden on L392. 

I think this method should only be responsible for returning the error and the caller methods (like `handleS3CredentialspecFile`) should be responsible for updating the terminal reason.

How about returning an `fmt.Errorf("error unmarshalling credentialspec domainless '%s': %w", err)` from here and not updating the terminal reason?

Same comment for L478 below.

**Review by arun-annamalai — APPROVED** [review]

File: agent/taskresource/credentialspec/credentialspec_linux_test.go (hunk 1 above)
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux_test.go:175]: I have a separate test that also covers the getmapping part, but this test it testing all the way from getting SSM info to the get mapping, I would remove it but I feel confident to have this assert ensuring in future nothing mess up the update mapping

File: agent/taskresource/credentialspec/credentialspec_linux.go (hunk 7 above)
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux.go]: updated accordingly

File: agent/taskresource/credentialspec/credentialspec_linux_test.go (hunk 1 above)
Comment (amogh09) [reply agent/taskresource/credentialspec/credentialspec_linux_test.go:175]: I see. Yeah I think we are mixing two test into one - 1) testing that `handleSSMCredentialspecFile` is able to call SSM client to download the data and parse it, and 2) `GetTargetMapping` is able to see updates made to `CredSpecMap` after  `handleSSMCredentialspecFile` has finished.

Maybe we can leave a comment explaining why the `GetTargetMapping` test is being performed in `TestHandleSSMDomainlessCredentialspecFile` as it is not clear just by looking at the test.

I still don’t see how `handleSSMCredentialspecFile` method is related to `CredSpecMap` though. Will leave this as a nit.

**Review by amogh09 — APPROVED** [review]

File: agent/taskresource/credentialspec/credentialspec_linux_test.go (hunk 1 above)
Comment (saikiranakula-amzn) [reply agent/taskresource/credentialspec/credentialspec_linux_test.go:175]: we will take a look at it to see if that can be removed. Thanks Amogh

**Review by mythri-garaga — APPROVED** [review]

**Pull request merged**